pkg/util/ranger/points.go

```diff
@@ -135,6 +136,47 @@ func rangePointEqualValueLess(a, b *point) bool {
 	return a.excl && !b.excl
 }
 
+func pointsConvertToSortKey(sctx sessionctx.Context, inputPs []*point, newTp *types.FieldType) ([]*point, error) {
+	if newTp.EvalType() != types.ETString ||
+		newTp.GetType() == mysql.TypeEnum ||
+		newTp.GetType() == mysql.TypeSet {
+		return inputPs, nil
+	}
+	ps := make([]*point, 0, len(inputPs))
+	for _, p := range inputPs {
+		np, err := pointConvertToSortKey(sctx, p, newTp, true)
+		if err != nil {
+			return nil, err
+		}
+		ps = append(ps, np)
+	}
+	return ps, nil
+}
+
+func pointConvertToSortKey(
+	sctx sessionctx.Context,
+	inputP *point,
+	newTp *types.FieldType,
+	trimTrailingSpace bool,
+) (*point, error) {
+	p, err := convertPoint(sctx, inputP, newTp)
+	if err != nil {
+		return nil, err
+	}
+	if p.value.Kind() != types.KindString || newTp.GetCollate() == charset.CollationBin {
+		return p, nil
+	}
+	sortKey := p.value.GetBytes()
+	if collate.NewCollationEnabled() {
+		if !trimTrailingSpace {
+			sortKey = collate.GetCollator(newTp.GetCollate()).KeyWithoutTrimRightSpace(string(hack.String(sortKey)))
+		} else {
+			sortKey = collate.GetCollator(newTp.GetCollate()).Key(string(hack.String(sortKey)))
+		}
+	}
+	return &point{value: types.NewBytesDatum(sortKey), excl: p.excl, start: p.start}, nil
+}
```

Comment:
Or
```
if !collate.NewCollationEnabled {
	return inputP
}
...
```
will be better?

Reply:
Updated. Merged this check into previous `if` branch.